From 8eafdb93fd551bfed7184bc7519f5d455811cefb Mon Sep 17 00:00:00 2001
From: Davide Beatrici <git@davidebeatrici.dev>
Date: Sat, 21 Feb 2026 05:33:36 +0100
Subject: [PATCH 2/3] __archive_read_register_bidder(): Allow
 ARCHIVE_STATE_HEADER, check if already registered

This allows formats to apply filters when they know what they need.

In the case of RPM, that's after parsing the (uncompressed) header.

Upstream-Status: Submitted [https://github.com/libarchive/libarchive/pull/2846]

Signed-off-by: Robert Yang <liezhi.yang@windriver.com>
---
 libarchive/archive_read.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/libarchive/archive_read.c b/libarchive/archive_read.c
index c9b9d598..bc6305ae 100644
--- a/libarchive/archive_read.c
+++ b/libarchive/archive_read.c
@@ -1232,13 +1232,18 @@ __archive_read_register_bidder(struct archive_read *a,
 	int i, number_slots;
 
 	archive_check_magic(&a->archive, ARCHIVE_READ_MAGIC,
-	    ARCHIVE_STATE_NEW, "__archive_read_register_bidder");
+	    ARCHIVE_STATE_NEW | ARCHIVE_STATE_HEADER,
+		"__archive_read_register_bidder");
 
 	number_slots = sizeof(a->bidders) / sizeof(a->bidders[0]);
 
 	for (i = 0; i < number_slots; i++) {
-		if (a->bidders[i].vtable != NULL)
+		if (a->bidders[i].vtable != NULL) {
+			if (a->bidders[i].vtable == vtable)
+				/* Already registered. */
+				return (ARCHIVE_OK);
 			continue;
+		}
 		memset(a->bidders + i, 0, sizeof(a->bidders[0]));
 		bidder = (a->bidders + i);
 		bidder->data = bidder_data;
-- 
2.49.0

